### PR TITLE
Fix focus loss from `Combobox.Input` when disabled option clicked

### PR DESCRIPTION
**Pull request by taro-28**
Fixed an issue with the `Combobox` component where clicking a disabled `Combobox.Option` would take focus away from `Combobox.Input`.

[sample codesandbox](https://codesandbox.io/p/sandbox/headless-ui-combobox-with-disabled-option-sn2ktl?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clsy8bjzk00063b83mjysamyk%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clsy8bjzk00023b83gkf19q9s%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clsy8bjzk00033b837ly238uh%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clsy8bjzk00053b83xqc7dm95%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clsy8bjzk00023b83gkf19q9s%2522%253A%257B%2522id%2522%253A%2522clsy8bjzk00023b83gkf19q9s%2522%252C%2522tabs%2522%253A%255B%255D%257D%252C%2522clsy8bjzk00053b83xqc7dm95%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clsy8bjzk00043b83gw4rpglo%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522clsy8bjzk00053b83xqc7dm95%2522%252C%2522activeTabId%2522%253A%2522clsy8bjzk00043b83gw4rpglo%2522%257D%252C%2522clsy8bjzk00033b837ly238uh%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522clsy8bjzk00033b837ly238uh%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

### CODE REVIEW

File: packages/@headlessui-react/src/components/combobox/combobox.tsx
Comment (taro-28) [inline packages/@headlessui-react/src/components/combobox/combobox.tsx:1]: Set `tabIndex=-1` for `Combobox.Option` when it is disabled.

Moved `data.inputRef.current?.focus({ preventScroll: true })` from `handleClick` to `handleFocus` and ensured it is called even when the option is disabled.

The reason for moving it into `handleFocus` is that the `onClick` event handler for disabled options is ignored by the `mergePropsAdvanced` function.

File: packages/@headlessui-react/src/test-utils/interactions.ts [hunk 1]
```diff
@@ -486,6 +488,14 @@ let focusableSelector = [
   )
   .join(',')
 
+let focusableSelectorWithNegativeTabindex = _focusableSelector
+  .map(
+    process.env.NODE_ENV === 'test'
+      ? (selector) => `${selector}:not([style*='display: none'])`
+      : (selector) => selector
+  )
+  .join(',')
```
Comment (taro-28) [inline packages/@headlessui-react/src/test-utils/interactions.ts:497]: An element with `tabIndex=-1` can be focused on through means other than keyboard navigation, such as clicking.

File: packages/@headlessui-vue/src/utils/render.ts [hunk 2]
```diff
@@ -218,16 +218,14 @@ function mergeProps(...listOfProps: Record<any, any>[]) {
...
-      // existing and new props, they don't just override therefore we have to
-      // explicitly nullify them.
-      Object.fromEntries(Object.keys(eventHandlers).map((eventName) => [eventName, undefined]))
-    )
+    for (let eventName in eventHandlers) {
+      // Prevent default events for `onClick`, `onMouseDown`, `onKeyDown`, etc.
+      if (/^(on(?:Click|Pointer|Mouse|Key)(?:Down|Up|Press)?)$/.test(eventName)) {
+        eventHandlers[eventName] = [(e: any) => e?.preventDefault?.()]
+      }
+    }
```
Comment (taro-28) [inline packages/@headlessui-vue/src/utils/render.ts:228]: Fixed it in the same way as the `mergePropsAdvanced` function in `@headlessui-react`.

https://github.com/tailwindlabs/headlessui/pull/2890

**Comment by RobinMalfait** [comment]
Hey!

We recently made some changes to how the `ComboboxInput`, `ComboboxButton` and `ComboboxOption` work behind the scenes. The short story is that we reworked the internals such that the focus doesn't leave the `ComboboxInput` at all. More info: https://github.com/tailwindlabs/headlessui/pull/3073

This means that in the latest insiders version clicking a disabled option should already work as expected.

Going to close this PR because it's not needed anymore, but I really appreciate your work on this!

Thanks!

**Pull request closed without merging**